File: ios/LanjingQuiz/Networking/ResultPageParser.swift
import Foundation

/// Port of the score / beatRate / rank regexes in server.js /api/exams/:id/submit.
enum ResultPageParser {

    static func parse(_ html: String) -> ExamResult {
        let ns = html as NSString
        let fullRange = NSRange(location: 0, length: ns.length)

        let score: String
        if let regex = try? NSRegularExpression(pattern: "class=\"score\"[^>]*>\\s*([\\d.]+)\\s*<"),
           let match = regex.firstMatch(in: html, range: fullRange) {
            score = ns.substring(with: match.range(at: 1))
        } else {
            score = "0"
        }

        var nums: [String] = []
        if let regex = try? NSRegularExpression(pattern: "exam-result-percentage[^>]*>\\s*(\\d+)") {
            nums = regex.matches(in: html, range: fullRange).map { ns.substring(with: $0.range(at: 1)) }
        }
        let beatRate = nums.first ?? "?"
        let rank = nums.count > 1 ? nums[1] : nums.first ?? "?"

        return ExamResult(score: score, beatRate: beatRate, rank: rank)
    }
}

File: apps/android/LanjingQuiz/src/main/java/com/qzh/lanjingquiz/Domain/ResultPageParser.kt
package com.qzh.lanjingquiz.Domain

import com.qzh.lanjingquiz.Network.ExamResult

/** 结果页解析 —— iOS ResultPageParser.swift 逐行移植(正则来自 apps/web/server.js)。 */
object ResultPageParser {

    fun parse(html: String): ExamResult {
        val score = Regex("""class="score"[^>]*>\s*([\d.]+)\s*<""").find(html)?.groupValues?.get(1) ?: "0"
        val nums = Regex("""exam-result-percentage[^>]*>\s*(\d+)""").findAll(html)
            .map { it.groupValues[1] }.toList()
        val beatRate = nums.firstOrNull() ?: "?"
        val rank = nums.getOrNull(1) ?: nums.firstOrNull() ?: "?"
        return ExamResult(score, beatRate, rank)
    }
}
